File: scripts/incremental_adjacency.py
from typing import Dict, List, Any

from scripts.adjacency_list import AdjacencyList


class IncrementalAdjacency:
    """
    Maintains and updates an adjacency list incrementally using change files.
    """

    def __init__(self):
        self.helper = AdjacencyList()

    def normalize(self, name: str) -> str:
        return self.helper.normalize_classname(name)

    def ensure_node(self, adj: Dict[str, List[str]], node: str):
        if node not in adj:
            adj[node] = []

    def add_edge(self, adj: Dict[str, List[str]], src: str, dst: str):
        self.ensure_node(adj, src)
        self.ensure_node(adj, dst)
        if dst not in adj[src]:
            adj[src].append(dst)
            adj[src].sort()

    def remove_edge(self, adj: Dict[str, List[str]], src: str, dst: str):
        if src in adj and dst in adj[src]:
            adj[src].remove(dst)

    def remove_node(self, adj: Dict[str, List[str]], node: str):
        if node in adj:
            del adj[node]
        # Remove any inbound references
        for n, neighbors in list(adj.items()):
            if node in neighbors:
                neighbors.remove(node)

    def apply_change(self, adj: Dict[str, List[str]], change_data: Dict[str, Any]) -> Dict[str, List[str]]:
        """
        Apply a single change JSON (map keyed by class) to the adjacency list.
        Uses delta.providersAdded/providersRemoved to adjust edges.
        Removes nodes when status.removed is true; ensures nodes on added.
        """
        for classname, details in change_data.items():
            n_class = self.normalize(classname)
            status = details.get("status", {})
            if status.get("removed"):
                self.remove_node(adj, n_class)
                continue

            # Ensure node exists (for added or modified)
            self.ensure_node(adj, n_class)

            delta = details.get("delta", {})
            for provider_str in delta.get("providersAdded", []):
                consumer = self.helper.extract_consumer_from_provider(provider_str)
                if consumer:
                    self.add_edge(adj, self.normalize(consumer), n_class)

            for provider_str in delta.get("providersRemoved", []):
                consumer = self.helper.extract_consumer_from_provider(provider_str)
                if consumer:
                    self.remove_edge(adj, self.normalize(consumer), n_class)

        return adj

File: scripts/IncrementalAdjacency.ts
// IncrementalAdjacency.ts
import { AdjacencyList } from './AdjacencyList';

export class IncrementalAdjacency {
  helper: AdjacencyList;

  constructor() {
    this.helper = new AdjacencyList();
  }

  normalize(name: string): string {
    return this.helper.normalizeClassname(name);
  }

  ensureNode(adj: Record<string, string[]>, node: string): void {
    if (!(node in adj)) {
      adj[node] = [];
    }
  }

  addEdge(adj: Record<string, string[]>, src: string, dst: string): void {
    this.ensureNode(adj, src);
    this.ensureNode(adj, dst);
    if (!adj[src].includes(dst)) {
      adj[src].push(dst);
      adj[src].sort();
    }
  }

  removeEdge(adj: Record<string, string[]>, src: string, dst: string): void {
    if (src in adj && adj[src].includes(dst)) {
      adj[src] = adj[src].filter(n => n !== dst);
    }
  }

  removeNode(adj: Record<string, string[]>, node: string): void {
    if (node in adj) {
      delete adj[node];
    }
    // Remove any inbound references
    for (const n in adj) {
      adj[n] = adj[n].filter(neighbor => neighbor !== node);
    }
  }

  applyChange(adj: Record<string, string[]>, changeData: Record<string, any>): Record<string, string[]> {
    for (const classname in changeData) {
      const details = changeData[classname];
      const nClass = this.normalize(classname);
      const status = details.status || {};
      if (status.removed) {
        this.removeNode(adj, nClass);
        continue;
      }

      // Ensure node exists (for added or modified)
      this.ensureNode(adj, nClass);

      const delta = details.delta || {};
      for (const providerStr of delta.providersAdded || []) {
        const consumer = this.helper.extractConsumerFromProvider(providerStr);
        if (consumer) {
          this.addEdge(adj, this.normalize(consumer), nClass);
        }
      }

      for (const providerStr of delta.providersRemoved || []) {
        const consumer = this.helper.extractConsumerFromProvider(providerStr);
        if (consumer) {
          this.removeEdge(adj, this.normalize(consumer), nClass);
        }
      }
    }
    return adj;
  }
}